### ISSUE TITLE
Different results in inference and validation?

### ISSUE THREAD

**Issue by trnikon** [ultralytics/ultralytics#4696]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am trying to plot the confusion matrix in a binary classification problem. I have 2 approaches, one using `results = model.val(...)` and `results.confusion_matrix.plot()` that plots as I would expect (about 0.96 accuracy) and in the other one I get the results using `results[0].probs.top1` and plot them using sklearn ConfusionMatrixDisplay. The problem is, in the second approach the predictions are all over the place, giving an accuracy of about 0.6!

Here is the relevant code for the second (wrong) approach:
```
    predictions = []
    targets = []
    transform = T.ToPILImage()

    # data_module.test_dataloader() returns a pytorch dataloader
    for images, labels in tqdm(data_module.test_dataloader()):
        targets.extend(labels.cpu().numpy())
        classes = []
        for img in images:
            results = model(transform(img))
            cls = results[0].probs.top1
            classes.append(cls)
        predictions.extend(classes)

    plot_cf_matrix(torch.Tensor(predictions), torch.Tensor(targets), labels=list(data_module.class_names.values()))
```
I have checked the predictions myself and indeed they are different than the targets. 
Any idea what might be wrong?

**Comment by glenn-jocher** [comment]
@trnikon thank you for providing a detailed explanation of your issue.

What you're experiencing could be due to a couple of factors. Firstly, the `model.val(...)` function and the `model(transform(img))` call might be working on different modes. `model.val(...)` works on evaluation mode (`model.eval()`) where certain functionalities like Dropout and BatchNorm are disabled to get consistent predictions. However, if your model is still in training mode during the `model(transform(img))` call, it could lead to variable results.

Additionally, the image transformations applied could also be different in both scenarios. The transformations applied to your validation images for the `model.val(...)` function might not be the same as the `T.ToPILImage()` transformation you are using for `model(transform(img))`. This could potentially lead to different results.

Another important aspect is how the targets are handled. Make sure the labels used by `model.val(...)` and those provided to `data_module.test_dataloader()` are consistent.

Lastly, I noticed you're taking `results[0].probs.top1` which gives the class with the highest predicted score. But for binary classification, it'd be more appropriate to use a threshold-based decision rather than taking the class with the highest probability. It might be worthwhile reconsidering this approach.

I hope this provides some direction in understanding the discrepancy. Double-checking these areas could potentially help explain the variance you're seeing.

**Comment by glenn-jocher** [comment]
Thank you for the follow-up questions, @trnikon. Let's take one at a time.

1. Regarding model evaluation mode, it's true that YOLO models often use the `.eval()` call to switch the model to evaluation mode. However, the object you're trying to call `.eval()` on seems to be a 'YOLO' object, not a PyTorch model. The error message indicates that `eval` is not a method available in the 'YOLO' object. For setting a YOLO model to evaluation mode, depending on your version of YOLO, you might need to use a method specific to YOLO.

2. About the image transformations, each model might use specific image preprocessing steps. Typically, these steps include some sort of normalization (to zero mean and unit variance) and resizing the image to the input size expected by the model. It's advisable to mimic these transformations manually, especially when exporting to ONNX as the model usually exports the weights only, not the preprocessing steps. You would need to check your specific model architecture for these details.

3. For the last point, it's great to hear that you've managed to synchronize the labels used in model validation and prediction, resulting in a more consistent outcome. It reiterates the fact that keeping the expectations and ground truth consistent across validation and prediction is crucial for receiving consistent results.

I hope I was able to clear some things up. Please feel free to ask if you have additional questions!

**Comment by thomaskleiven** [comment]
Similar problem as discussed in issue #4536. 

@trnikon did you find a solution to this problem?

**Comment by glenn-jocher** [comment]
@thomaskleiven hello, 

It appears you're facing a similar issue as discussed in #4536. The issue in #4536 was partly resolved by ensuring consistency between the labels used during the model's validation phase and the labels provided for predictions. 

In terms of setting the model into evaluation mode, it's likely that the `YOLO` object being used didn't directly support this, as it's probably not a PyTorch model instance. You might need to look into a YOLO-specific way of setting the model to evaluation mode if it's not done automatically during inference. 

Regarding image transformations, the transformations applied during the validation phase need to be replicated when generating predictions. This process may meet your needs better than using generic transformations, as model-specific transformations often include changes like normalization and resizing to a certain input size.

I hope this provides some guidance to help you out with your problem. If you have further questions or if something isn't clear, please feel free to ask.
